rust/gsc/src/read_job.rs

```diff
@@ -1,12 +1,24 @@
+use crate::types::{Dir, File, ReadJobPayload, ReadJobResponse, Symlink};
 use axum::Json;
-use entity::{job, output_dir, output_file, output_symlink};
+use chrono::Utc;
+use entity::{job, job_uses, output_dir, output_file, output_symlink};
 use sea_orm::{
-    ColumnTrait, DatabaseConnection, DbErr, EntityTrait, ModelTrait, QueryFilter, TransactionTrait,
+    ActiveModelTrait, ActiveValue, ColumnTrait, DatabaseConnection, DbErr, EntityTrait, ModelTrait,
+    QueryFilter, TransactionTrait,
 };
 use std::sync::Arc;
 use tracing;
 
-use crate::types::{Dir, File, ReadJobPayload, ReadJobResponse, Symlink};
+#[tracing::instrument]
+async fn record_use(job_id: i32, conn: Arc<DatabaseConnection>) {
```

Comment:
I think functions like this are better  written as extentions (I think in Rust they are called Impls?). Basically add this as a member function of  `Job`. Then you can do `matching_job.record_use(conn)`

Ideally we should avoid passing  around parts of a model and instead just act directly on the model

Reply:
Ok, I have to use a trait to do this but this is a not-uncommon pattern to use (this is why you see traits that end in Ext a lot)

Reply:
So ...actually... this is difficult but it is doable....but like wait a second.

1) You can't define an impl outside of the crate that the struct was defined in
2) You *can* define a trait however and this is a not-uncommon pattern. You'll often see traits that end in `Ext` that do exactly this. but....
3) You can't add async functions inside of traits because `async fn foo(...) -> T` desugars into `fn foo(...) -> impl Future<Output = T>` which works fine in an `impl` because no associated type is needed but that's an implicit associated type if its in a trait buuut.....
4) There's a trait called `async-trait` that handles this for you and makes a bespoke associated type for each method aaaaaaand...
5) Tyler and the rest of the compiler team are working on getting async traits in by default so that you don't need a macro for it buuuuut....
6) I don't think it will be stable for a while

We could do the async-trait thing and wait for a proper language feature OR we could avoid all that hassle and just pass a `job::Model&` instead of part of it....however either way we have to consider something else:

There's a second issue to contend with:
1) `i32` implements copy but `job::Model` does not (nor should it). It should however implement `clone` but `clone` could be extremely expensive because of stdout/stderr
2) I could however move this into an arc (cheap enough) and then call this in the spawned task as `record_use(matching_job_arc.as_ref(), new_conn)`

I think making it a regular function instead of a method and going with option 2) following that is the best compromise

Funny how rust complicates things like this huh?

Reply:
I see. hmmm. Rust does complicate this a bit.

I think I want to avoid polluting stuff with arc just yet and the copy would be too expensive. 

This is easily fixed in the future so lets just leave this as is until things have shaken  out a bit in `async-trait` and also we more generally have a feeling for what rust-in-wake looks like

Reply:
Also reading Tyler's blog post it turns out that I was slightly wrong on point 4 there. async-trait actully creates a `Pin<Box<dyn Future<Output = T>>>` for each method which is super-duper slow. That probably isn't too big of a cost however given that this is a spawned write task.